### PR TITLE
Add short lived memory cache for tokens acquired by DefaultAzureCredential

### PR DESCRIPTION
**Pull request by pavel-purma**
Cherry pick to update release 8.0

### CODE REVIEW

File: src/Microsoft.DotNet.Build.Tasks.Feed/src/AssetPublisherFactory.cs [hunk 1]
```diff
@@ -38,22 +40,36 @@ public virtual IAssetPublisher CreateAssetPublisher(TargetFeedConfig feedConfig,
...
+                            ExcludeVisualStudioCredential = true,
+                            ExcludeAzureDeveloperCliCredential = true,
+                            ExcludeInteractiveBrowserCredential = true,
+                            ManagedIdentityClientId = mi,
+                            CredentialProcessTimeout = TimeSpan.FromSeconds(60.0)
+                        }
+                    )
+                );
+            });
+            return tokenCredential;
```
Comment (premun) [inline src/Microsoft.DotNet.Build.Tasks.Feed/src/AssetPublisherFactory.cs:71]: Could be this as well:

```suggestion
            return _tokenCredentialsPerManagedIdentity.GetOrAdd(managedIdentityClientId, static (mi) =>
                new TokenCredentialShortCache(
                    new DefaultAzureCredential(
                        new DefaultAzureCredentialOptions
                        {
                            ExcludeVisualStudioCodeCredential = true,
                            ExcludeVisualStudioCredential = true,
                            ExcludeAzureDeveloperCliCredential = true,
                            ExcludeInteractiveBrowserCredential = true,
                            ManagedIdentityClientId = mi,
                            CredentialProcessTimeout = TimeSpan.FromSeconds(60.0)
                        }));
```

File: src/Microsoft.DotNet.Build.Tasks.Feed/src/common/TokenCredentialShortCache.cs [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+    public override AccessToken GetToken(TokenRequestContext requestContext, CancellationToken cancellationToken)
+    {
+        return GetTokenAsync(requestContext, cancellationToken).Result;
+    }
+
+    public override async ValueTask<AccessToken> GetTokenAsync(TokenRequestContext requestContext, CancellationToken cancellationToken)
+    {
+        CacheKey cacheKey = new CacheKey
+        {
+            Scopes = String.Join(":", requestContext.Scopes),
```
Comment (premun) [inline src/Microsoft.DotNet.Build.Tasks.Feed/src/common/TokenCredentialShortCache.cs:37]:
```suggestion
            Scopes = string.Join(':', requestContext.Scopes),
```

File: src/Microsoft.DotNet.Build.Tasks.Feed/src/common/TokenCredentialShortCache.cs [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+            // go over all the cache items and remove the ones that are eligible to remove
+            _tokenCache.Keys.ToList().ForEach(key =>
+            {
+                if (_tokenCache.TryGetValue(key, out CachedToken ct))
+                {
+                    if (ct.EligibleToRemove)
+                    {
+                        _tokenCache.TryRemove(key, out _);
+                    }
+                }
```
Comment (premun) [inline src/Microsoft.DotNet.Build.Tasks.Feed/src/common/TokenCredentialShortCache.cs:64]:
```suggestion
                if (_tokenCache.TryGetValue(key, out CachedToken ct) && ct.EligibleToRemove)
                {
                    _tokenCache.TryRemove(key, out _);
                }
```

**Review by premun — APPROVED** [review]

File: src/Microsoft.DotNet.Build.Tasks.Feed/src/AssetPublisherFactory.cs (hunk 1 above)
Comment (pavel-purma) [reply src/Microsoft.DotNet.Build.Tasks.Feed/src/AssetPublisherFactory.cs:71]: I will update it to this

File: src/Microsoft.DotNet.Build.Tasks.Feed/src/common/TokenCredentialShortCache.cs [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+            Claims = requestContext.Claims,
+            TenantId = requestContext.TenantId,
+            IsCaeEnabled = requestContext.IsCaeEnabled
+        };
+
+        bool doCleanUpCache = false;
+
+        CachedToken cachedToken = _tokenCache.GetOrAdd(cacheKey, _ => new CachedToken());
+        var token = await cachedToken.GetToken(requestContext, () => {
+            // initiate cleanup of cache only when we are going to get any fresh token
```
[the remark this answers is not in the document]
Comment (pavel-purma) [reply src/Microsoft.DotNet.Build.Tasks.Feed/src/common/TokenCredentialShortCache.cs:47]: It is mentioned at clean up action itself. I have added some details there

**Review by premun — APPROVED** [review]
What were the problems with `ICacheMemory`? Seems like we have a lot of code for things that are there out of the box (like the expiration, clean-up..)

**Review by premun — APPROVED** [review]

**Comment by pavel-purma** [comment]
Memory Cache would work perfectly if we lock by semaphore against all cache items, solve just one at a time. But then I refactored it to use semaphore for locking just for one of the cache key. That slightly collide with logic of cache which should contain only some value for period of time, controlled by cache and remain not mutated during time. In my code I also use CacheKey record which leverage GetHashCode and Equality implementation of records. It seems like cache did not work for me the same index key addressing as dictionary works.

**Pull request merged**